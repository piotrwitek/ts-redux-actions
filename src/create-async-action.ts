import {
  TypeConstant,
  ActionCreatorBuilder,
  ActionBuilder,
} from './type-helpers';
import { createAction } from './create-action';

export function throwInvalidAsyncActionArgument(argPosition: number): never {
  throw new Error(
    `Argument ${argPosition} is invalid, it should be an action type of "string | symbol" or a tuple of "[string | symbol, Function, Function?]"`
  );
}

type AsyncActionHandler<
  TType extends TypeConstant,
  TArgs extends any[],
  TPayloadMeta
> = [TArgs] extends [never]
  ? ActionCreatorBuilder<
      TType,
      unknown extends TPayloadMeta
        ? any
        : [TPayloadMeta] extends [[infer T, any]]
        ? T
        : TPayloadMeta,
      unknown extends TPayloadMeta
        ? undefined
        : [TPayloadMeta] extends [[any, infer T]]
        ? T
        : undefined
    >
  : (
      ...args: TArgs
    ) => ActionBuilder<
      TType,
      [TPayloadMeta] extends [[infer T, any]] ? T : TPayloadMeta,
      [TPayloadMeta] extends [[any, infer T]] ? T : undefined
    >;

interface AsyncAction<
  TType1 extends TypeConstant,
  TPayload1 extends any,
  TMeta1 extends any,
  TArgs1 extends any[],
  TType2 extends TypeConstant,
  TPayload2 extends any,
  TMeta2 extends any,
  TArgs2 extends any[],
  TType3 extends TypeConstant,
  TPayload3 extends any,
  TMeta3 extends any,
  TArgs3 extends any[],
  TType4 extends TypeConstant,
  TPayload4 extends any,
  TMeta4 extends any,
  TArgs4 extends any[]
> {
  // tslint:disable-next-line: callable-types
  <
    TPayloadMeta1 extends
      | TPayload1
      | [TPayload1, TMeta1] = TMeta1 extends undefined
      ? TPayload1
      : [TPayload1, TMeta1],
    TPayloadMeta2 extends
      | TPayload2
      | [TPayload2, TMeta2] = TMeta2 extends undefined
      ? TPayload2
      : [TPayload2, TMeta2],
    TPayloadMeta3 extends
      | TPayload3
      | [TPayload3, TMeta3] = TMeta3 extends undefined
      ? TPayload3
      : [TPayload3, TMeta3],
    TPayloadMeta4 extends
      | TPayload4
      | [TPayload4, TMeta4] = TMeta3 extends undefined
      ? TPayload4
      : [TPayload4, TMeta4]
  >(): [TType4] extends [never]
    ? {
        request: AsyncActionHandler<TType1, TArgs1, TPayloadMeta1>;
        success: AsyncActionHandler<TType2, TArgs2, TPayloadMeta2>;
        failure: AsyncActionHandler<TType3, TArgs3, TPayloadMeta3>;
      }
    : {
        request: AsyncActionHandler<TType1, TArgs1, TPayloadMeta1>;
        success: AsyncActionHandler<TType2, TArgs2, TPayloadMeta2>;
        failure: AsyncActionHandler<TType3, TArgs3, TPayloadMeta3>;
        cancel: AsyncActionHandler<TType4, TArgs4, TPayloadMeta4>;
      };
}

export function createAsyncAction<
  TType1 extends TypeConstant,
  TType2 extends TypeConstant,
  TType3 extends TypeConstant,
  TType4 extends TypeConstant = never
>(
  requestArg: TType1,
  successArg: TType2,
  failureArg: TType3,
  cancelArg?: TType4
): AsyncAction<
  TType1,
  unknown,
  unknown,
  never,
  TType2,
  unknown,
  unknown,
  never,
  TType3,
  unknown,
  unknown,
  never,
  TType4,
  unknown,
  unknown,
  never
>;

export function createAsyncAction<
  TType1 extends TypeConstant,
  TType2 extends TypeConstant,
  TType3 extends TypeConstant,
  TType4 extends TypeConstant = never,
  TPayloadCreator1 extends
    | ((...args: TArgs1) => TPayload1)
    | undefined = undefined,
  TPayloadCreator2 extends
    | ((...args: TArgs2) => TPayload2)
    | undefined = undefined,
  TPayloadCreator3 extends
    | ((...args: TArgs3) => TPayload3)
    | undefined = undefined,
  TPayloadCreator4 extends
    | ((...args: TArgs4) => TPayload4)
    | undefined = undefined,
  TMetaCreator1 extends ((...args: TArgs1) => TMeta1) | undefined = undefined,
  TMetaCreator2 extends ((...args: TArgs2) => TMeta2) | undefined = undefined,
  TMetaCreator3 extends ((...args: TArgs3) => TMeta3) | undefined = undefined,
  TMetaCreator4 extends ((...args: TArgs4) => TMeta4) | undefined = undefined,
  TPayload1 extends any = TPayloadCreator1 extends ((...args: any[]) => infer T)
    ? T
    : undefined,
  TMeta1 extends any = TMetaCreator1 extends ((...args: any[]) => infer T)
    ? T
    : undefined,
  TPayload2 extends any = TPayloadCreator2 extends ((...args: any[]) => infer T)
    ? T
    : undefined,
  TMeta2 extends any = TMetaCreator2 extends ((...args: any[]) => infer T)
    ? T
    : undefined,
  TPayload3 extends any = TPayloadCreator3 extends ((...args: any[]) => infer T)
    ? T
    : undefined,
  TMeta3 extends any = TMetaCreator3 extends ((...args: any[]) => infer T)
    ? T
    : undefined,
  TPayload4 extends any = TPayloadCreator4 extends ((...args: any[]) => infer T)
    ? T
    : undefined,
  TMeta4 extends any = TMetaCreator4 extends ((...args: any[]) => infer T)
    ? T
    : undefined,
  TArgs1 extends any[] = TPayloadCreator1 extends ((...args: infer T) => any)
    ? T
    : TMetaCreator1 extends ((...args: infer T) => any)
    ? T
    : never,
  TArgs2 extends any[] = TPayloadCreator2 extends ((...args: infer T) => any)
    ? T
    : TMetaCreator2 extends ((...args: infer T) => any)
    ? T
    : never,
  TArgs3 extends any[] = TPayloadCreator3 extends ((...args: infer T) => any)
    ? T
    : TMetaCreator3 extends ((...args: infer T) => any)
    ? T
    : never,
  TArgs4 extends any[] = TPayloadCreator4 extends ((...args: infer T) => any)
    ? T
    : TMetaCreator4 extends ((...args: infer T) => any)
    ? T
    : never
>(
  requestArg:
    | TType1
    | [TType1, TPayloadCreator1]
    | [TType1, TPayloadCreator1, TMetaCreator1],
  successArg:
    | TType2
    | [TType2, TPayloadCreator2]
    | [TType2, TPayloadCreator2, TMetaCreator2],
  failureArg:
    | TType3
    | [TType3, TPayloadCreator3]
    | [TType3, TPayloadCreator3, TMetaCreator3],
  cancelArg?:
    | TType4
    | [TType4, TPayloadCreator4]
    | [TType4, TPayloadCreator4, TMetaCreator4]
): AsyncAction<
  TType1,
  TPayload1,
  TMeta1,
  TArgs1,
  TType2,
  TPayload2,
  TMeta2,
  TArgs2,
  TType3,
  TPayload3,
  TMeta3,
  TArgs3,
  TType4,
  TPayload4,
  TMeta4,
  TArgs4
>;

/**
 * @description create an async action-creator object that contains `request`, `success` and `failure` actions as props
 */
export function createAsyncAction<
  TType1 extends TypeConstant,
  TType2 extends TypeConstant,
  TType3 extends TypeConstant,
  TType4 extends TypeConstant = never,
  TPayloadCreator1 extends
    | ((...args: TArgs1) => TPayload1)
    | undefined = undefined,
  TPayloadCreator2 extends
    | ((...args: TArgs2) => TPayload2)
    | undefined = undefined,
  TPayloadCreator3 extends
    | ((...args: TArgs3) => TPayload3)
    | undefined = undefined,
  TPayloadCreator4 extends
    | ((...args: TArgs4) => TPayload4)
    | undefined = undefined,
  TMetaCreator1 extends ((...args: TArgs1) => TMeta1) | undefined = undefined,
  TMetaCreator2 extends ((...args: TArgs2) => TMeta2) | undefined = undefined,
  TMetaCreator3 extends ((...args: TArgs3) => TMeta3) | undefined = undefined,
  TMetaCreator4 extends ((...args: TArgs4) => TMeta4) | undefined = undefined,
  TPayload1 extends any = TPayloadCreator1 extends ((...args: any[]) => infer T)
    ? T
    : undefined,
  TMeta1 extends any = TMetaCreator1 extends ((...args: any[]) => infer T)
    ? T
    : undefined,
  TPayload2 extends any = TPayloadCreator2 extends ((...args: any[]) => infer T)
    ? T
    : undefined,
  TMeta2 extends any = TMetaCreator2 extends ((...args: any[]) => infer T)
    ? T
    : undefined,
  TPayload3 extends any = TPayloadCreator3 extends ((...args: any[]) => infer T)
    ? T
    : undefined,
  TMeta3 extends any = TMetaCreator3 extends ((...args: any[]) => infer T)
    ? T
    : undefined,
  TPayload4 extends any = TPayloadCreator4 extends ((...args: any[]) => infer T)
    ? T
    : undefined,
  TMeta4 extends any = TMetaCreator4 extends ((...args: any[]) => infer T)
    ? T
    : undefined,
  TArgs1 extends any[] = TPayloadCreator1 extends ((...args: infer T) => any)
    ? T
    : TMetaCreator1 extends ((...args: infer T) => any)
    ? T
    : never,
  TArgs2 extends any[] = TPayloadCreator2 extends ((...args: infer T) => any)
    ? T
    : TMetaCreator2 extends ((...args: infer T) => any)
    ? T
    : never,
  TArgs3 extends any[] = TPayloadCreator3 extends ((...args: infer T) => any)
    ? T
    : TMetaCreator3 extends ((...args: infer T) => any)
    ? T
    : never,
  TArgs4 extends any[] = TPayloadCreator4 extends ((...args: infer T) => any)
    ? T
    : TMetaCreator4 extends ((...args: infer T) => any)
    ? T
    : never
>(
  requestArg:
    | TType1
    | [TType1, TPayloadCreator1]
    | [TType1, TPayloadCreator1, TMetaCreator1],
  successArg:
    | TType2
    | [TType2, TPayloadCreator2]
    | [TType2, TPayloadCreator2, TMetaCreator2],
  failureArg:
    | TType3
    | [TType3, TPayloadCreator3]
    | [TType3, TPayloadCreator3, TMetaCreator3],
  cancelArg?:
    | TType4
    | [TType4, TPayloadCreator4]
    | [TType4, TPayloadCreator4, TMetaCreator4]
): AsyncAction<
  TType1,
  TPayload1,
  TMeta1,
  TArgs1,
  TType2,
  TPayload2,
  TMeta2,
  TArgs2,
  TType3,
  TPayload3,
  TMeta3,
  TArgs3,
  TType4,
  TPayload4,
  TMeta4,
  TArgs4
> {
  const constructor = (<
    TP1 = undefined,
    TM1 = undefined,
    TP2 = undefined,
    TM2 = undefined,
    TP3 = undefined,
    TM3 = undefined,
    TP4 = undefined,
    TM4 = undefined
  >() => {
    const results = [requestArg, successArg, failureArg, cancelArg].map(
      (arg, index) => {
        if (Array.isArray(arg)) {
          return createAction(arg[0], arg[1] as any, arg[2] as any)();
        } else if (typeof arg === 'string' || typeof arg === 'symbol') {
          return createAction(arg as string)();
        } else if (index < 3) {
          throwInvalidAsyncActionArgument(index);
        }
      }
    );

    const [request, success, failure, cancel] = results;

    return {
      request,
      success,
      failure,
      cancel,
    };
  }) as AsyncAction<
    TType1,
    TPayload1,
    TMeta1,
    TArgs1,
    TType2,
    TPayload2,
    TMeta2,
    TArgs2,
    TType3,
    TPayload3,
    TMeta3,
    TArgs3,
    TType4,
    TPayload4,
    TMeta4,
    TArgs4
  >;

  return constructor;
}
