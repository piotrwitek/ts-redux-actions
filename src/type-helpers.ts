/**
 * PUBLIC API
 */

/**
 * @desc Type representing Type Constant
 */
export type TypeConstant = string;

/**
 * @desc Type representing Generic Action
 */
export type Action<TType extends TypeConstant = TypeConstant> = {
  type: TType;
};

/**
 * @desc Type representing Generic ActionCreator
 */
export type ActionCreator<TType extends TypeConstant> = (
  ...args: any[]
) => Action<TType>;

/**
 * @desc Type representing Generic Reducer
 */
export type Reducer<TState, TAction extends Action> = (
  state: TState | undefined,
  action: TAction
) => TState;

/**
 * @desc Action without Payload
 */
export type EmptyAction<TType extends TypeConstant> = {
  type: TType;
};

/**
 * @desc Action with only Payload
 */
export type PayloadAction<TType extends TypeConstant, TPayload> = {
  type: TType;
  payload: TPayload;
};

/**
 * @desc Action with only Meta
 */
export type MetaAction<TType extends TypeConstant, TMeta> = {
  type: TType;
  meta: TMeta;
};

/**
 * @desc Action with both Payload and Meta
 */
export type PayloadMetaAction<TType extends TypeConstant, TPayload, TMeta> = {
  type: TType;
  payload: TPayload;
  meta: TMeta;
};

/**
 * @desc Action Creator producing EmptyAction
 */
export type EmptyAC<TType extends TypeConstant> = () => EmptyAction<TType>;

/**
 * @desc Action Creator producing PayloadAction
 */
export type PayloadAC<TType extends TypeConstant, TPayload> = (
  payload: TPayload
) => PayloadAction<TType, TPayload>;

/**
 * @desc Action Creator producing PayloadMetaAction
 */
export type PayloadMetaAC<TType extends TypeConstant, TPayload, TMeta> = (
  payload: TPayload,
  meta: TMeta
) => PayloadMetaAction<TType, TPayload, TMeta>;

/**
 * @desc Type representing type getter on Action Creator instance
 */
export interface TypeMeta<TType extends TypeConstant> {
  getType?: () => TType;
}

/**
 * @desc Infers Action union-type from action-creator map object
 */
export type ActionType<
  TActionCreatorOrMap extends any
> = TActionCreatorOrMap extends ActionCreator<TypeConstant>
  ? ReturnType<TActionCreatorOrMap>
  : TActionCreatorOrMap extends Record<any, any>
  ? {
      [K in keyof TActionCreatorOrMap]: ActionType<TActionCreatorOrMap[K]>
    }[keyof TActionCreatorOrMap]
  : TActionCreatorOrMap extends infer R // TODO: should be just never but compiler yell with circularly references itself error
  ? never
  : never;

/**
 * @desc Infers State object from reducer map object
 */
export type StateType<
  TReducerOrMap extends any
> = TReducerOrMap extends Reducer<any, any>
  ? ReturnType<TReducerOrMap>
  : TReducerOrMap extends Record<any, any>
  ? { [K in keyof TReducerOrMap]: StateType<TReducerOrMap[K]> }
  : never;

/**
 * INTERNAL API
 */

/** @private */
export type ActionBuilderConstructor<
  TType extends TypeConstant,
  TPayload extends any = undefined,
  TMeta extends any = undefined
> = [TMeta] extends [undefined]
  ? [TPayload] extends [undefined]
    ? unknown extends TPayload
      ? PayloadAC<TType, TPayload>
      : unknown extends TMeta
      ? PayloadMetaAC<TType, TPayload, TMeta>
      : EmptyAC<TType>
    : PayloadAC<TType, TPayload>
  : PayloadMetaAC<TType, TPayload, TMeta>;

/** @private */
export type ActionBuilderMap<
  TType extends TypeConstant,
  TActionProps extends any,
  TPayloadArg extends any = undefined,
  TMetaArg extends any = undefined
> = [TMetaArg] extends [undefined]
  ? [TPayloadArg] extends [undefined]
    ? () => { type: TType } & TActionProps
    : (payload: TPayloadArg) => { type: TType } & TActionProps
  : (payload: TPayloadArg, meta: TMetaArg) => { type: TType } & TActionProps;
