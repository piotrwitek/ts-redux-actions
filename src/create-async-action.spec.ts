import * as TH from './type-helpers';
import { AsyncActionCreatorBuilder } from './type-helpers';
import { createAsyncAction } from './create-async-action';

type User = { firstName: string; lastName: string };

// @dts-jest:group async action without cancel
{
  const fetchUsersAsync = createAsyncAction(
    'FETCH_USERS_REQUEST',
    'FETCH_USERS_SUCCESS',
    'FETCH_USERS_FAILURE'
  )<undefined, User[], Error>();

  // @dts-jest:pass:snap
  fetchUsersAsync.request(); /* => {
    type: 'FETCH_USERS_REQUEST'
  } */

  // @dts-jest:pass:snap
  fetchUsersAsync.success([
    { firstName: 'Piotr', lastName: 'Witek' },
  ]); /* => {
    type: 'FETCH_USERS_SUCCESS', payload: [{ firstName: 'Piotr', lastName: 'Witek' }]
  } */

  // @dts-jest:pass:snap
  fetchUsersAsync.failure(
    Error('reason')
  ); /* => {
    type: 'FETCH_USERS_FAILURE', payload: Error('reason')
  } */

  // @dts-jest:pass:snap
  fetchUsersAsync.cancel;

  const fn = (
    a: AsyncActionCreatorBuilder<
      ['FETCH_USERS_REQUEST', undefined],
      ['FETCH_USERS_SUCCESS', User[]],
      ['FETCH_USERS_FAILURE', Error]
    >
  ) => {
    a.request;
    a.success;
    a.failure;
    a.cancel;
    return a;
  };
  // @dts-jest:pass:snap
  fn(fetchUsersAsync);
}

// @dts-jest:group async action with any type
{
  const fetchUsersAsync = createAsyncAction(
    'FETCH_USERS_REQUEST',
    'FETCH_USERS_SUCCESS',
    'FETCH_USERS_FAILURE'
  )<any, any[], any>();

  // @dts-jest:pass:snap
  fetchUsersAsync.request(
    1
  ); /* => {
    type: 'FETCH_USERS_REQUEST', payload: 1,
  } */

  // @dts-jest:pass:snap
  fetchUsersAsync.success([
    1,
  ]); /* => {
    type: 'FETCH_USERS_SUCCESS', payload: [1],
  } */

  // @dts-jest:pass:snap
  fetchUsersAsync.failure(
    1
  ); /* => {
    type: 'FETCH_USERS_FAILURE', payload: 1,
  } */

  // @dts-jest:pass:snap
  fetchUsersAsync.cancel;

  const fn = (
    a: AsyncActionCreatorBuilder<
      ['FETCH_USERS_REQUEST', any],
      ['FETCH_USERS_SUCCESS', any[]],
      ['FETCH_USERS_FAILURE', any]
    >
  ) => {
    a.request;
    a.success;
    a.failure;
    a.cancel;
    return a;
  };
  // @dts-jest:pass:snap
  fn(fetchUsersAsync);
}

// @dts-jest:group async action with cancel
{
  const fetchUsersAsync = createAsyncAction(
    'FETCH_USERS_REQUEST',
    'FETCH_USERS_SUCCESS',
    'FETCH_USERS_FAILURE',
    'FETCH_USERS_CANCEL'
  )<undefined, User[], Error, string>();

  // @dts-jest:pass:snap
  fetchUsersAsync.request(); /* => {
    type: 'FETCH_USERS_REQUEST'
  } */

  // @dts-jest:pass:snap
  fetchUsersAsync.success([
    { firstName: 'Piotr', lastName: 'Witek' },
  ]); /* => {
    type: 'FETCH_USERS_SUCCESS', payload: [{ firstName: 'Piotr', lastName: 'Witek' }]
  } */

  // @dts-jest:pass:snap
  fetchUsersAsync.failure(
    Error('reason')
  ); /* => {
    type: 'FETCH_USERS_FAILURE', payload: Error('reason')
  } */

  // @dts-jest:pass:snap
  fetchUsersAsync.cancel(
    'reason'
  ); /* => {
    type: 'FETCH_USERS_CANCEL', payload: 'reason'
  } */

  const fn = (
    a: AsyncActionCreatorBuilder<
      ['FETCH_USERS_REQUEST', undefined],
      ['FETCH_USERS_SUCCESS', User[]],
      ['FETCH_USERS_FAILURE', Error],
      ['FETCH_USERS_CANCEL', string]
    >
  ) => {
    {
      a.request;
      a.success;
      a.failure;
      a.cancel;
      return a;
    }
  };
  // @dts-jest:pass:snap
  fn(fetchUsersAsync);
}

// @dts-jest:group async action with meta
{
  const fetchUsersAsync = createAsyncAction(
    'FETCH_USERS_REQUEST',
    'FETCH_USERS_SUCCESS',
    'FETCH_USERS_FAILURE'
  )<[undefined, number], User[], [Error, number]>();

  // @dts-jest:pass:snap
  fetchUsersAsync.request(
    undefined,
    111
  ); /* => {
    type: 'FETCH_USERS_REQUEST', meta: 111
  } */

  // @dts-jest:pass:snap
  fetchUsersAsync.success([
    { firstName: 'Piotr', lastName: 'Witek' },
  ]); /* => {
    type: 'FETCH_USERS_SUCCESS', payload: [{ firstName: 'Piotr', lastName: 'Witek' }]
  } */

  // @dts-jest:pass:snap
  fetchUsersAsync.failure(
    Error('reason'),
    111
  ); /* => {
    type: 'FETCH_USERS_FAILURE', payload: Error('reason'), meta: 111
  } */

  // @dts-jest:pass:snap
  fetchUsersAsync.cancel;

  const fn = (
    a: AsyncActionCreatorBuilder<
      ['FETCH_USERS_REQUEST', [undefined, number]],
      ['FETCH_USERS_SUCCESS', User[]],
      ['FETCH_USERS_FAILURE', [Error, number]]
    >
  ) => {
    a.request;
    a.success;
    a.failure;
    a.cancel;
    return a;
  };
  // @dts-jest:pass:snap
  fn(fetchUsersAsync);
}

// @dts-jest:group async action with meta with cancel
{
  const fetchUsersAsync = createAsyncAction(
    'FETCH_USERS_REQUEST',
    'FETCH_USERS_SUCCESS',
    'FETCH_USERS_FAILURE',
    'FETCH_USERS_CANCEL'
  )<[undefined, number], User[], [Error, number], string>();

  // @dts-jest:pass:snap
  fetchUsersAsync.request(
    undefined,
    111
  ); /* => {
    type: 'FETCH_USERS_REQUEST', meta: 111
  } */

  // @dts-jest:pass:snap
  fetchUsersAsync.success([
    { firstName: 'Piotr', lastName: 'Witek' },
  ]); /* => {
    type: 'FETCH_USERS_SUCCESS', payload: [{ firstName: 'Piotr', lastName: 'Witek' }]
  } */

  // @dts-jest:pass:snap
  fetchUsersAsync.failure(
    Error('reason'),
    111
  ); /* => {
    type: 'FETCH_USERS_FAILURE', payload: Error('reason'), meta: 111
  } */

  // @dts-jest:pass:snap
  fetchUsersAsync.cancel(
    'reason'
  ); /* => {
    type: 'FETCH_USERS_CANCEL', payload: 'reason'
  } */

  const fn = (
    a: AsyncActionCreatorBuilder<
      ['FETCH_USERS_REQUEST', [undefined, number]],
      ['FETCH_USERS_SUCCESS', User[]],
      ['FETCH_USERS_FAILURE', [Error, number]],
      ['FETCH_USERS_CANCEL', string]
    >
  ) => {
    a.request;
    a.success;
    a.failure;
    a.cancel;
    return a;
  };
  // @dts-jest:pass:snap
  fn(fetchUsersAsync);
}

// @dts-jest:group async action with mappers
{
  const fetchUserMappers = createAsyncAction(
    'FETCH_USER_REQUEST',
    [
      'FETCH_USER_SUCCESS',
      ({ firstName, lastName }: User) => `${firstName} ${lastName}`,
    ],
    [
      'FETCH_USER_FAILURE',
      (error: Error, meta: number) => error,
      (error: Error, meta: number) => meta,
    ]
  )();

  // @dts-jest:pass:snap
  fetchUserMappers.request(); /* => {
    type: 'FETCH_USER_REQUEST'
  } */

  // @dts-jest:pass:snap
  fetchUserMappers.success({
    firstName: 'Piotr',
    lastName: 'Witek',
  }); /* => {
    type: 'FETCH_USER_SUCCESS',
    payload: 'Piotr Witek',
  } */

  // @dts-jest:pass:snap
  fetchUserMappers.failure(
    Error('reason'),
    111
  ); /* => {
    type: 'FETCH_USER_FAILURE', payload: Error('reason'), meta: 111
  } */

  // @dts-jest:pass:snap
  fetchUserMappers.cancel;

  const fn = (
    a: AsyncActionCreatorBuilder<
      ['FETCH_USER_REQUEST', undefined],
      ['FETCH_USER_SUCCESS', [User], string],
      ['FETCH_USER_FAILURE', [Error, number], [Error, number]]
    >
  ) => {
    a.request;
    a.success;
    a.failure;
    a.cancel;
    return a;
  };
  // @dts-jest:pass:snap
  fn(fetchUserMappers);
}

// @dts-jest:group async action with mappers with cancel
{
  const fetchUserMappers = createAsyncAction(
    'FETCH_USER_REQUEST',
    [
      'FETCH_USER_SUCCESS',
      ({ firstName, lastName }: User) => `${firstName} ${lastName}`,
    ],
    [
      'FETCH_USER_FAILURE',
      (error: Error, meta: number) => error,
      (error: Error, meta: number) => meta,
    ],
    ['FETCH_USER_CANCEL', undefined, (meta: number) => meta]
  )();

  // @dts-jest:pass:snap
  fetchUserMappers.request(); /* => {
    type: 'FETCH_USER_REQUEST'
  } */

  // @dts-jest:pass:snap
  fetchUserMappers.success({
    firstName: 'Piotr',
    lastName: 'Witek',
  }); /* => {
    type: 'FETCH_USER_SUCCESS',
    payload: 'Piotr Witek',
  } */

  // @dts-jest:pass:snap
  fetchUserMappers.failure(
    Error('reason'),
    111
  ); /* => {
    type: 'FETCH_USER_FAILURE', payload: Error('reason'), meta: 111
  } */

  // @dts-jest:pass:snap
  fetchUserMappers.cancel(
    111
  ); /* => {
    type: 'FETCH_USER_CANCEL', meta: 111
  } */

  const fn = (
    a: AsyncActionCreatorBuilder<
      ['FETCH_USER_REQUEST', undefined],
      ['FETCH_USER_SUCCESS', [User], string],
      ['FETCH_USER_FAILURE', [Error, number], [Error, number]],
      ['FETCH_USER_CANCEL', [number], [undefined, number]]
    >
  ) => {
    a.request;
    a.success;
    a.failure;
    a.cancel;
    return a;
  };
  // @dts-jest:pass:snap
  fn(fetchUserMappers);
}
