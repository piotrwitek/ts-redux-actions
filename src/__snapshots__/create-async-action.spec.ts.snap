// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`async action with any type fetchUsers.failure(
    1
  ) (type) should match snapshot 1`] = `"T.PayloadAction<\\"FETCH_USERS_FAILURE\\", any>"`;

exports[`async action with any type fetchUsers.request(
    1
  ) (type) should match snapshot 1`] = `"T.PayloadAction<\\"FETCH_USERS_REQUEST\\", any>"`;

exports[`async action with any type fetchUsers.success([
    1,
  ]) (type) should match snapshot 1`] = `"T.PayloadAction<\\"FETCH_USERS_SUCCESS\\", any[]>"`;

exports[`async action with undefined type fetchUsers.failure(
    Error('reason')
  ) (type) should match snapshot 1`] = `"T.PayloadAction<\\"FETCH_USERS_FAILURE\\", Error>"`;

exports[`async action with undefined type fetchUsers.request() (type) should match snapshot 1`] = `"T.EmptyAction<\\"FETCH_USERS_REQUEST\\">"`;

exports[`async action with undefined type fetchUsers.success([
    { firstName: 'Piotr', lastName: 'Witek' },
  ]) (type) should match snapshot 1`] = `"T.PayloadAction<\\"FETCH_USERS_SUCCESS\\", User[]>"`;
